index.js

```diff
@@ -938,6 +933,145 @@ DB.prototype._visitTrie = function (key, path, node, heads, halt, visited, cb, t
   }
 }
 
+DB.prototype.createHistoryStream = function (opts) {
+  opts = opts || {}
+
+  var self = this
+
+  var stream = new Readable({objectMode: true})
+  stream._read = noop
+
+  function cb (err) {
+    stream.emit('error', err)
+  }
+
+  var nodes = []
+  var seqs = []
+  var atFront = false  // whether the live stream has reached real-time
+
+  // Populate 'nodes' and 'seq' with first entry in each writer
+  initTraversal(function (err, theNodes, theSeqs) {
+    if (err) return cb(err)
+    nodes = theNodes
+    seqs = theSeqs
+    traverseNode()
+  })
+
+  function traverseNode () {
+    var next = nodes.reduce(
+      function (a, b) {
+        if (!a && !b) return null
+        else if (!a) return b
+        else if (!b) return a
+        else if (a.seq <= b.clock[a.feed]) return a
+        else return b
+      }, null)
+
+    // Avoid ending the stream with a null value if opts.live
+    if (!opts.live || next !== null) stream.push(next)
+
+    // No new nodes to traverse
+    if (!next) {
+      // Stream has ended
+      if (atFront || !opts.live) return
+
+      // Stop traversal; set up event listeners; wait
+      atFront = true
+      self.on('_writer', onWriter)
+      self.on('append', onAppend)
```

Comment:
don't you need to remove the onAppend, onWriter listeners after on of the fires to avoid this firing twice?

Reply:
Otherwise two traverseNode "threads" might be running in parallel if two 'append' events happen inbetween onAppend firing and it triggering resumeTraversal async